app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/FahrplanFragment.kt

```diff
@@ -0,0 +1,744 @@
+        super.onCreateOptionsMenu(menu, inflater)
+        inflater.inflate(R.menu.fahrplan_menu, menu)
+    }
+
+    override fun onOptionsItemSelected(menuItem: MenuItem): Boolean {
+        return if (menuItem.itemId == R.id.menu_item_refresh) {
+            // TODO Replace MainActivity reference with viewModel.
+            val activity = requireActivity() as MainActivity
+            activity.fetchFahrplan()
+            true
+        } else {
+            super.onOptionsItemSelected(menuItem)
+        }
+    }
+
+    private fun updateMenuItems() {
+        // Toggles the icon for "add/delete favorite" or "add/delete alarm".
+        // Triggers SessionDetailsFragment.onPrepareOptionsMenu to be called
+        requireActivity().invalidateOptionsMenu()
+    }
+
+    override fun onCreateContextMenu(menu: ContextMenu, view: View, menuInfo: ContextMenuInfo?) {
+        super.onCreateContextMenu(menu, view, menuInfo)
+        contextMenuView = view
+        val session = view.tag as Session
+        if (session.highlight) {
+            menu.add(0, CONTEXT_MENU_ITEM_ID_FAVORITES, 0, getString(R.string.menu_item_title_unflag_as_favorite))
+        } else {
+            menu.add(0, CONTEXT_MENU_ITEM_ID_FAVORITES, 0, getString(R.string.menu_item_title_flag_as_favorite))
+        }
+        if (session.hasAlarm) {
+            menu.add(0, CONTEXT_MENU_ITEM_ID_DELETE_ALARM, 2, getString(R.string.menu_item_title_delete_alarm))
+        } else {
+            menu.add(0, CONTEXT_MENU_ITEM_ID_SET_ALARM, 1, getString(R.string.menu_item_title_set_alarm))
+        }
+        menu.add(0, CONTEXT_MENU_ITEM_ID_ADD_TO_CALENDAR, 3, getString(R.string.menu_item_title_add_to_calendar))
+        if (BuildConfig.ENABLE_CHAOSFLIX_EXPORT) {
+            val share = menu.addSubMenu(0, CONTEXT_MENU_ITEM_ID_SHARE, 4, getString(R.string.menu_item_title_share_session))
+            share.add(0, CONTEXT_MENU_ITEM_ID_SHARE_TEXT, 5, getString(R.string.menu_item_title_share_session_text))
+            share.add(0, CONTEXT_MENU_ITEM_ID_SHARE_JSON, 6, getString(R.string.menu_item_title_share_session_json))
+        } else {
+            menu.add(0, CONTEXT_MENU_ITEM_ID_SHARE, 4, getString(R.string.menu_item_title_share_session))
+        }
+    }
+
+    private fun getSessionView(session: Session): View? {
+        val verticalScrollView = requireView().findViewById<NestedScrollView?>(R.id.verticalScrollView)
+            ?: return null
+        return verticalScrollView.findViewWithTag(session)
+    }
+
+    private fun refreshViews() {
+        if (scheduleData == null) {
+            return
+        }
+        for (session in scheduleData!!.allSessions) {
+            setBell(session)
+            val sessionView = getSessionView(session)
+            if (sessionView != null) {
+                sessionViewDrawer.setSessionBackground(session, sessionView)
+                SessionViewDrawer.setSessionTextColor(session, sessionView)
+            }
+        }
+    }
```

Comment:
This one almost slip...as with the other you can get rid of the return by negating the condition. However since you already mentioned you will refactor this in a follow up...please just resolve it as you read it.

Reply:
OK :+1: The `refreshViews` function will be gone soon.